packages/@lwc/template-compiler/src/codegen/optimize.ts

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import * as astring from 'astring';
+import { walk } from 'estree-walker';
+import * as t from '../shared/estree';
+
+/**
+ * Given a template function, extract all static objects/arrays (e.g. `{ key : 1 }`)
+ * and return them as a list of `const` declarations. Also return the modified function
+ * that is now referencing those declarations.
+ *
+ * Example input:
+ *
+ * ```
+ * function tmpl() {
+ *   return {
+ *     foo: dynamic(),
+ *     bar: { static: true },
+ *     baz: { really: { static: ['yep', 'totally', 'static' ] } }
+ *   };
+ * }
+ * ```
+ *
+ * Example output:
+ *
+ * ```
+ * const stc0 = { static: true };
+ * const stc1 = { really: { static: ['yep', 'totally', 'static' ] } };
+ * function tmpl() {
+ *   return {
+ *     foo: dynamic(),
+ *     bar: stc0,
+ *     baz: stc1
+ *   };
+ * }
+ * ```
+ */
+export function optimizeStaticExpressions(
+    templateFn: t.FunctionDeclaration
+): Array<t.FunctionDeclaration | t.VariableDeclaration> {
+    const result: Array<t.FunctionDeclaration | t.VariableDeclaration> = [];
+    const keysToVariableNames = new Map();
+
+    let identifierCount = 0;
+
+    // Return true if this node is an object/array that is fully static
+    function isStaticObjectOrArray(node: t.BaseNode): boolean {
+        if (t.isObjectExpression(node)) {
+            return node.properties.every((prop) => {
+                return (
+                    t.isProperty(prop) &&
+                    !prop.computed &&
+                    !prop.method &&
+                    !prop.shorthand &&
+                    (t.isLiteral(prop.value) || isStaticObjectOrArray(prop.value))
+                );
+            });
+        } else if (t.isArrayExpression(node)) {
+            return node.elements.every((element) => {
+                return (
+                    element !== null &&
+                    t.isExpression(element) &&
+                    (t.isLiteral(element) || isStaticObjectOrArray(element))
+                );
+            });
+        }
+        return false;
+    }
+
+    function extractStaticVariable(node: t.BaseNode): t.BaseNode {
```

Comment:
Instead of accepting any AST node type, I would restrict the signature further, removing the type guard in the variable declarator and moving the type guard to the tree walker enter method. Having the type guard right after invoking `isStaticObjectOrArray` would make more sense when reading the code.

```suggestion
    function extractStaticVariable(node: t.Expression): t.Identifier {
```


```js
            if (isStaticObjectOrArray(node)) {
                const newNode = extractStaticVariable(node as t.Expression);
                this.replace(newNode);
                this.skip();
            }
```